### ISSUE TITLE
Problems with showing the Notebook Status in the Controller and the Jupyter Webapp

### ISSUE THREAD

**Issue by kimwnasptd** [kubeflow/kubeflow#3820]
Right now, there are some problems with how we expose the Notebook's underlying Status in the Notebook CR. The current behavior in the Notebook Controller is this:

1. We expose only the `contaierStatus[0]` of the underlying Pod in the Notebook CR's Status
2. We update the `conditions` in the status of the Notebook CR from the uuu Pod, but only when the `containerStatuses` [are not zero](https://github.com/kubeflow/kubeflow/blob/master/components/notebook-controller/pkg/controller/notebook/notebook_controller.go#L264).

But due to these, the following problems arise:

1. The `containerStatuses` in the Pod gets set only once the Pod is scheduled. This results in the following:
    * If the StatefulSet can't create the Pod (e.x. missing ServiceAccount), then the Notebook CR's status (`containerState`) is not updated / initialized.
    * If the StatefulSet creates the Pod, but the Pod is not Schedulable (ex. missing PVC, CPU limits), then `containerStatus` doesn't get set both in Pod's and Notebook CR's status. Instead the error info is shown in the Pod's `conditions` with Type `Unscheduled` and the relevant `reason` and `message`.

1. Since we have Istio, the `containerState[0]` might be the State from the Sidecar Container. If that's the case then the UI shows the `containerState` of the Sidecar. 
  
1. WebApp looks at the `containerState` in the Notebook CR to deduce the status. So if the Pod is not schedulable, then the Status doesn't get updated and the webapp shows generic `Scheduling the Pod`. Also, as noted above, it might be showing the status of another Container.

To fix these we will need to make changes both to the jupyter webapp and in the notebook controller.

In the Controller, we need to:
* Update the Notebook CR's `conditions` even when the statefuleset can't create the Pod. For this, the controller will also need to reconcile the `Events` k8s objects and update the CR's `conditions` when the Pod is not created. This is tracked [here](https://github.com/kubeflow/kubeflow/issues/3772).
* Update the Notebook CR's `conditions` with the Pod's `conditions`, even when the Pod is not Scheduled. This way we expose information for scheduling problems like missing PVCs or insufficient CPU/Memory.
* Expose all the `containerStatuses` of the Pod's Containers and not only the first one. This means that we will need to modify the [status](https://github.com/kubeflow/kubeflow/blob/master/components/notebook-controller/pkg/apis/notebook/v1alpha1/notebook_types.go#L39) in Notebook CRD.

In the Webapp, we need to:
* Look at the CR's `conditions` if the `containerStatuses` are zero. This way we will surface to the user problems with the Pod's scheduling.
* To determine whether to show a {green check, spinner, warning, error} sign, the webapp will check all of the `containerStatuses` and:
    * if: all of the Containers are `running`, then it will show a green check.
    * else if: at least one container is in `waiting` state and with `reason: ImagePullBackOff`, then it will show an error sign.
    * else if: the `reason` is either `ContainerCreating` or `PodInitializing`, then it will show a spinner
    * else: it will show a warning sign 

cc @jlewi @lluunn

**Comment by jlewi** [comment]
Related: #3405 jupyter-web-app should read notebook conditions

Related: #3772 Lots of discussion about how to surface statefulset events in the jupyter CR and web app

**Comment by jlewi** [comment]
@kimwnasptd should we close this as a duplicate of #3637? or possibly #3772?

Based on the various discussion I believe the plan (#3772) is to have the notebook controller replay appropriate events and then the jupyter web app can surface those events (#3637)?

**Comment by kimwnasptd** [comment]
Yes, since we will be using Events with the work for #3637 and #3772 we can drop the proposed implementation in this issue.

Although the problem 2 I described, the use of `containerStatus[0]` within the controller, remains. So we might still be exposing the status of a sidecar container instead of the Notebook Container. But either way lets close this issue and I'll open a new issue to only keep track of the mentioned problematic behavior.

**Comment by jlewi** [comment]
Duplicate of #3637
